File: Solutions/897. Increasing Order Search Tree/897. Increasing Order Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
 //Recursive
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        return foo(root,nullptr);
    }
    TreeNode* foo(TreeNode* root,TreeNode* tail){
        if(!root)return tail;
        TreeNode* res=foo(root->left,root);
        root->left=nullptr;
        root->right=foo(root->right,tail);
        return res;
    }
};

//Iterative
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        stack<TreeNode*> st;
        //point to the leftmostchild
        TreeNode* dummyleftupside=new TreeNode();
		TreeNode* dummymove=dummyleftupside;
        while(root||st.size()) {
			if(root) {
				st.push(root);
				root=root->left;
			}
			else {
				TreeNode* current=st.top();
                st.pop();				
				//tricky part,check if there is righthchild 
				root=current->right;
				current->left=nullptr;
				dummymove->right=current;
				dummymove=dummymove->right;	
			}
		}
		return dummyleftupside->right;
    }
};

File: Solutions/897. Increasing Order Search Tree/897. Increasing Order Search Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 //recursive
func increasingBST(root *TreeNode) *TreeNode {
    return increasing(root,nil)
}

func increasing(cur,tail *TreeNode) *TreeNode {
    if cur==nil{
        return tail
    }
    res:=increasing(cur.Left,cur)
    cur.Left=nil
    cur.Right=increasing(cur.Right,tail)
    return res
}


//iterative
func increasingBST(root *TreeNode) *TreeNode {
    var stk []*TreeNode
    dummyleft:=new(TreeNode)
    dummy:=dummyleft
    for root!=nil || len(stk)>0{
        if root!=nil{
            stk=append(stk,root)
            root=root.Left
        }else{
            cur:=stk[len(stk)-1]
            stk=stk[:len(stk)-1]
            root=cur.Right
            cur.Left=nil
            dummy.Right=cur
            dummy=dummy.Right
        }
    }
    return dummyleft.Right
}